### ISSUE TITLE
Error message’s hint pointing entirely at the wrong place.

### ISSUE THREAD

**Issue by steffahn** [rust-lang/rust#77677]
Reduced example:

```rust
trait TraitF {
    fn f();
}
trait TraitG {
    fn g();
}

struct Foo<T>(T);

impl<T> Foo<T> {
    fn foo() {}
    fn bar() {
        struct Bar<F, T>(F, T);
        impl<F, T> TraitF for Bar<F, T>
        where
            F: FnOnce(T),
        {
            fn f() {}
        }
        impl<F> TraitG for Bar<F>
        where
            F: FnOnce(T),
        {
            fn g() {}
        }
    }
}

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=466bc6b30842ff1e63f943bfce94b199))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0401]: can't use generic parameters from outer function
  --> src/lib.rs:22:23
   |
10 | impl<T> Foo<T> {
   |      - type parameter from outer function
...
18 |             fn f() {}
   |                - try adding a local generic parameter in this method instead
...
22 |             F: FnOnce(T),
   |                       ^ use of generic parameter from outer function

error[E0107]: wrong number of type arguments: expected 2, found 1
  --> src/lib.rs:20:28
   |
20 |         impl<F> TraitG for Bar<F>
   |                            ^^^^^^ expected 2 type arguments

error: aborting due to 2 previous errors

Some errors have detailed explanations: E0107, E0401.
For more information about an error, try `rustc --explain E0107`.
error: could not compile `playground`.

To learn more, run the command again with --verbose.

```

This came up during refactoring, while I was adding the parameter `T` to `Bar` (intentionally named the same as the outer `T` because they are going to be instantiated to the same type anyways). This kind of error message (the first one, `E0401`) had me puzzled for some time before figuring out that it was nonsensical. I’m referring to the `try adding a local generic parameter in this method instead` part. The real code of course was _quite a bit_ longer, making the situation _actually_ confusing, in particular:
* The `try adding ...` hint points to the wrong place entirely. It _should_ suggest adding a parameter to the `impl TraitG` (even though such a suggestion might not always make sense, idk).
* The `type parameter from outer function` pointer is also somewhat confusing, since when `foo`’s body is a few lines and `bar` is _super long and nested_, it is quite nontrivial to even figure out that “`outer function`” refers to `bar()`. (As in: it is easy to forget that you’re even inside of a function body at all if you’ve been focusing on this `Bar` struct and it’s `impl`s for a while.) Also... what is the distinction between the words `function` and `method` here? (My original code has `bar(self)` but still called it “`outer function`”.)

<hr>

@rustbot modify labels: A-diagnostics, T-compiler, D-confusing, C-bug.

**Comment by Enselic** [comment]
(I'm exploring the depths of the Rust issue tracker.) The errors given has changed to the following. Do you think it is still buggy/confusing?

```
Compiling playground v0.0.1 (/playground)
error[[E0401]](https://doc.rust-lang.org/stable/error_codes/E0401.html): can't use generic parameters from outer function
  --> src/lib.rs:22:23
   |
10 | impl<T> Foo<T> {
   |      - type parameter from outer function
...
20 |         impl<F> TraitG for Bar<F>
   |              - help: try using a local generic parameter instead: `T,`
21 |         where
22 |             F: FnOnce(T),
   |                       ^ use of generic parameter from outer function

error[[E0107]](https://doc.rust-lang.org/stable/error_codes/E0107.html): struct takes 2 generic arguments but 1 generic argument was supplied
  --> src/lib.rs:20:28
   |
20 |         impl<F> TraitG for Bar<F>
   |                            ^^^ - supplied 1 generic argument
   |                            |
   |                            expected 2 generic arguments
   |
note: struct defined here, with 2 generic parameters: `F`, `T`
  --> src/lib.rs:13:16
   |
13 |         struct Bar<F, T>(F, T);
   |                ^^^ -  -
help: add missing generic argument
   |
20 |         impl<F> TraitG for Bar<F, T>
   |                                 +++

Some errors have detailed explanations: E0107, E0401.
For more information about an error, try `rustc --explain E0107`.
error: could not compile `playground` (lib) due to 2 previous errors
```

**Comment by steffahn** [comment]
Thanks for noticing this @Enselic.

It appears like the bug was “accidentally” fixed in #99249

I wonder if the code that originally did something wrong for figuring out the source location still exists for any other purposes, or if it’s gone entirely. I would ask for feedback from @cjgillot (:wave:), but I don’t know how much one can remember one year after the PR, and as long as there are no known issues (and this is a diagnostics problem to begin with, only), it’s likely not worth spending any time investigating.

I think I will not keep this issue open for merely the suggestion to add a mention of the actual outer function `bar`; I can imagine it’d be nicer if the error message would point out `bar` for example by mentioning its name
```
error[[E0401]](https://doc.rust-lang.org/stable/error_codes/E0401.html): can't use generic parameters from outer function
  --> src/lib.rs:22:23
   |
10 | impl<T> Foo<T> {
   |      - type parameter from outer function `bar`
...
20 |         impl<F> TraitG for Bar<F>
   |              - help: try using a local generic parameter instead: `T,`
21 |         where
22 |             F: FnOnce(T),
   |                       ^ use of generic parameter from outer function
```
or perhaps even pointing to it? (Of course only in cases like this where the parameter is from the `impl` block, which might be somewhat distant from where the function `bar` itself begins. I’ll consider at a later time whether to open a new enhancement/papercut issue on that point.
